src/Symfony/Component/String/Utf8String.php

```diff
@@ -0,0 +1,360 @@
+
+    private static $transliterators = [];
+
+    public function __construct(string $string = '')
+    {
+        if ('' !== $string && !preg_match('//u', $string)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        $this->string = $string;
+    }
+
+    public static function fromCodePoints(int ...$codes): self
+    {
+        $string = '';
+        foreach ($codes as $code) {
+            if (0x80 > $code %= 0x200000) {
+                $string .= \chr($code);
+            } elseif (0x800 > $code) {
+                $string .= \chr(0xC0 | $code >> 6).\chr(0x80 | $code & 0x3F);
+            } elseif (0x10000 > $code) {
+                $string .= \chr(0xE0 | $code >> 12).\chr(0x80 | $code >> 6 & 0x3F).\chr(0x80 | $code & 0x3F);
+            } else {
+                $string .= \chr(0xF0 | $code >> 18).\chr(0x80 | $code >> 12 & 0x3F).\chr(0x80 | $code >> 6 & 0x3F).\chr(0x80 | $code & 0x3F);
+            }
+        }
+
+        $str = new static();
+        $str->string = $string;
+
+        return $str;
+    }
+
+    public function append(string ...$suffix): parent
+    {
+        $str = clone $this;
+        $str->string .= 1 >= \count($suffix) ? $suffix[0] ?? '' : implode('', $suffix);
+
+        if (!preg_match('//u', $str->string)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        return $str;
+    }
+
+    /**
+     * Generic UTF-8 to ASCII transliteration.
+     *
+     * Install the intl extension for best results.
+     *
+     * @param string[] $rules See "*-Latin" rules from Transliterator::listIDs()
+     */
+    public function ascii(array $rules = []): self
+    {
+        $str = clone $this;
+        $s = $str->string;
+        $str->string = '';
+        $step = 0;
+
+        if (\function_exists('transliterator_transliterate')) {
+            $rules[] = 'Any-Latin';
+            $rules[] = 'Latin-ASCII';
+        }
+
+        while (\strlen($s) !== $i = strspn($s, self::ASCII)) {
+            if (0 !== $i) {
+                $str->string .= substr($s, 0, $i);
+                $s = substr($s, $i);
+            }
+
+            if (1 === ++$step) {
+                if (!normalizer_is_normalized($s, self::NFKC)) {
+                    $s = normalizer_normalize($s, self::NFKC);
+                }
+            } elseif (2 === $step) {
+                $s = str_replace(self::TRANSLIT_FROM, self::TRANSLIT_TO, $s);
+            } elseif (3 === $step && null === $rule = array_shift($rules)) {
+                $s = preg_replace('/\p{Mn}++/u', '', normalizer_normalize($s, self::NFD));
+            } elseif (3 === $step) {
+                $step = 2;
+
+                if (\in_array($rule, ['de', 'DE', 'de-ascii', 'de-ASCII', 'DE-ASCII'], true)) {
```

Comment:
If the `de-ASCII` rule is available via `Intl` the `Transliterator` should be used instead of this manual replacement, because `Intl` supports contextual conversions (like “Äpfel” to “Aepfel” and “ÄPFEL” to “AEPFEL”).

Reply:
Thanks also, I fixed the mapping for `de-ascii`, it matches the CLDR definition now.